lib/logstash/inputs/kafka.rb

```diff
@@ -20,168 +21,89 @@
-          event = @kafka_client_queue.pop
-          queue_event(event, logstash_queue)
-        end
-      end
-
-      until @kafka_client_queue.empty?
-        queue_event(@kafka_client_queue.pop,logstash_queue)
-      end
-
-      @logger.info('Done running kafka input')
-    rescue => e
-      @logger.warn('kafka client threw exception, restarting',
-                   :exception => e)
-      Stud.stoppable_sleep(Float(@consumer_restart_sleep_ms) * 1 / 1000) { stop? }
-      retry if !stop?
-    end
+    @runner_consumers = num_threads.times.map { || new_consumer }
+    @runner_threads = @runner_consumers.map { |consumer| thread_runner(logstash_queue, consumer) }
+    @runner_threads.each { |t| t.join }
   end # def run
 
   public
   def stop
-    @consumer_group.shutdown if @consumer_group.running?
+    @runner_consumers.each { |c| c.wakeup }
   end
 
   private
-  def create_consumer_group(options)
-    Kafka::Group.new(options)
+  def thread_runner(logstash_queue, consumer)
+    Thread.new do
+      begin
+        consumer.subscribe(topics);
+        while !stop?
+          records = consumer.poll(timeout_ms);
+          for record in records do
+            break if !max_messages.nil? && @message_count >= max_messages
+            @message_count = @message_count.next
+            @codec.decode(record.value.to_s) do |event|
+              logstash_queue << event
+            end
+          end
+        end
+      rescue org.apache.kafka.common.errors.WakeupException => e
+        raise e if !stop?
+      ensure
+        consumer.close
+      end
+    end
   end
 
   private
-  def queue_event(message_and_metadata, output_queue)
-    begin
-      @codec.decode("#{message_and_metadata.message}") do |event|
-        decorate(event)
-        if @decorate_events
-          event['kafka'] = {'msg_size' => message_and_metadata.message.size,
-                            'topic' => message_and_metadata.topic,
-                            'consumer_group' => @group_id,
-                            'partition' => message_and_metadata.partition,
-                            'key' => message_and_metadata.key}
-        end
-        output_queue << event
-      end # @codec.decode
-    rescue => e # parse or event creation error
-      @logger.error('Failed to create event', :message => "#{message_and_metadata.message}", :exception => e,
-                    :backtrace => e.backtrace)
-    end # begin
-  end # def queue_event
+  def new_consumer 
+    props = java.util.Properties.new
+    props.put("bootstrap.servers", bootstrap_servers)
+    props.put("group.id", group_id);
+    props.put("enable.auto.commit", enable_auto_commit);
+    props.put("auto.commit.interval.ms", auto_commit_interval_ms);
+    props.put("session.timeout.ms", session_timeout_ms);
+    props.put("key.deserializer", key_deserializer);
+    props.put("value.deserializer", value_deserializer);
+    org.apache.kafka.clients.consumer.KafkaConsumer.new(props);
```

Comment:
We also need to pass all the configs here, does not look like the full list we expose above